go/vt/topo/shard_test.go

```diff
@@ -140,69 +140,148 @@ func TestUpdateSourcePrimaryDeniedTables(t *testing.T) {
-			DeniedTables: []string{"t1", "t2"},
+			name:       "re-add existing cell, different tables, should fail",
+			tabletType: topodatapb.TabletType_RDONLY,
+			cells:      []string{"first"},
+			tables:     []string{"t3"},
+			wantError:  "trying to use two different sets of denied tables for shard",
 		},
-	}) {
-		t.Fatalf("one cell removal from all failed: %v", si)
+		{
+			name:       "add all cells, see cell list grow to all",
+			tabletType: topodatapb.TabletType_RDONLY,
+			cells:      []string{"first", "second", "third"},
+			tables:     []string{"t1", "t2"},
+			wantTabletControl: &topodatapb.Shard_TabletControl{
+				TabletType:   topodatapb.TabletType_RDONLY,
+				Cells:        []string{"first", "second", "third"},
+				DeniedTables: []string{"t1", "t2"},
+			},
+		},
+		{
+			name:       "remove one cell",
+			tabletType: topodatapb.TabletType_RDONLY,
+			cells:      []string{"second"},
+			remove:     true,
+			tables:     []string{"t1", "t2"},
+			wantTabletControl: &topodatapb.Shard_TabletControl{
+				TabletType:   topodatapb.TabletType_RDONLY,
+				Cells:        []string{"first", "third"},
+				DeniedTables: []string{"t1", "t2"},
+			},
+		},
+		{
+			name:       "add replica tablet type",
+			tabletType: topodatapb.TabletType_REPLICA,
+			cells:      []string{"first"},
+			tables:     []string{"t1", "t2"},
+			wantTabletControl: &topodatapb.Shard_TabletControl{
+				TabletType:   topodatapb.TabletType_REPLICA,
+				Cells:        []string{"first"},
+				DeniedTables: []string{"t1", "t2"},
+			},
+		},
+		{
+			name:       "confirm rdonly still stays the same, after replica was added",
+			tabletType: topodatapb.TabletType_RDONLY,
+			wantTabletControl: &topodatapb.Shard_TabletControl{
+				TabletType:   topodatapb.TabletType_RDONLY,
+				Cells:        []string{"first", "third"},
+				DeniedTables: []string{"t1", "t2"},
+			},
+		},
+		{
+			name:       "remove rdonly entry",
+			tabletType: topodatapb.TabletType_RDONLY,
+			cells:      []string{"first", "third"},
+			remove:     true,
+			tables:     []string{"t1", "t2"},
+		},
+		{
+			name:       "remove replica entry",
+			tabletType: topodatapb.TabletType_REPLICA,
+			cells:      []string{"first", "third"},
+			remove:     true,
+			tables:     []string{"t1", "t2"},
+		},
+	}
+
+	for _, tcase := range testCases {
+		t.Run(tcase.name, func(t *testing.T) {
+			if tcase.ctx == nil {
+				tcase.ctx = ctxWithLock
+			}
+			var err error
+			if tcase.tables != nil || tcase.cells != nil {
+				err = si.UpdateDeniedTables(tcase.ctx, tcase.tabletType, tcase.cells, tcase.remove, tcase.tables)
+			}
+			if tcase.wantError != "" {
+				require.Error(t, err)
+				require.Contains(t, err.Error(), tcase.wantError)
+				return
+			}
+			require.NoError(t, err)
+			if tcase.wantTabletControl == nil {
+				require.Nil(t, si.GetTabletControl(tcase.tabletType))
+			} else {
+				require.Truef(t, reflect.DeepEqual(tcase.wantTabletControl, si.GetTabletControl(tcase.tabletType)),
```

Comment:
We should be able to use `require.Equal` or `require.EqualValues` here rather than `require.True` with `DeepEqual`.